src/attributes/diagnostics.md

```diff
@@ -83,6 +85,56 @@ pub mod m3 {
 > [command-line][rustc-lint-cli], and also supports [setting
 > caps][rustc-lint-caps] on the lints that are reported.
 
+All lint attributes support an additional `reason` parameter, to give context why
+a certain attribute was added. This reason will be displayed as part of the lint
+message, if the lint is emitted at the defined level.
+
+```rust
+use std::path::PathBuf;
+
+pub fn get_path() -> PathBuf {
+    #[allow(unused_mut, reason = "this is only modified on some platforms")]
+    let mut file_name = PathBuf::from("git");
+
+    #[cfg(target_os = "windows")]
+    file_name.set_extension("exe");
+
+    file_name
+}
+```
+
+### Lint expectations
+
+With the `#[expect]` attributes lints can be expected in a certain scope. If
+this expectation is not fulfilled a new warning is emitted to the user. The
+lint levels can be overridden with other lint attributes as usual.
+
+```rust
+#[warn(missing_docs)]
+pub mod m2{
+    #[expect(missing_docs)]
+    pub mod nested {
+        // This missing documentation fulfills the expectation above
+        pub fn undocumented_one() -> i32 { 1 }
+
+        // Missing documentation signals a warning here, despite the expectation
+        // above. This emission would not fulfill the expectation
+        #[warn(missing_docs)]
+        pub fn undocumented_two() -> i32 { 2 }
```

Comment:
Oh, I think I was a bit confused by the absence of docs on the modules.  I see that this does prevent it from being fulfilled.  This seems a little confusing, per my comment below.

Reply:
Good call, I fixed the example. I thought about using another lint, but decided against it, since this example was already used and might therefore be simpler to understand :)